#26 Implement getCurrentSingleSignOnAccount() livecycle aware

stefan-niedermann opened the pull request:
- Add `CurrentSingleSignOnAccountLiveData` class which registers a `OnSharedPreferencesChanged` listener in a lifecycle aware way (It basically does only emit new events when the `Activity` / `Fragment` is not `paused` or `stopped`
- Add `SingleAccountHelper#getCurrentSingleSignOnAccount$` which returns a `CurrentSingleSignOnAccountLiveData` instance. The trailing `$` is a convention borrowed from `rxjs`[<sup>[1]</sup>](https://angular.io/guide/rx-library#naming-conventions-for-observables) to signalize that this is something observable
- `@Deprecate`ed `SingleAccountHelper#(un)registerSharedPreferenceChangeListener` methods, which are not lifecycle aware and potential memory leaks when used wrong
- Add sample usage to sample app

Example usage from 3rd party app:

```java
SingleAccountHelper.getCurrentSingleSignOnAccount$(this).observe(this, ssoAccount -> {
    Log.i(TAG, "New SingleSignOnAccount set: " + ssoAccount.name);
});
```

Signed-off-by: Stefan Niedermann <info@niedermann.it>

David-Development commented:
This looks great! I haven't personally worked with Live data before but definetely a great addition and perfect use case! I think we should probably mention it in the docs, eh?  Not sure if devs will find it if it isn't documented in the Readme..?

stefan-niedermann commented:
The PR is nust to demonstrate the general idea - of course we can add some docs and discuss the API before merging this. 🙂

stefan-niedermann commented:
I have added minimal documentation in the `README.md` and in the sample app :slightly_smiling_face:

stefan-niedermann commented:
Would love to get a review @AndyScherzinger @David-Development :slightly_smiling_face:

AndyScherzinger commented on lib/src/main/java/com/nextcloud/android/sso/helper/SingleSignOnAccountLiveData.java:
```diff
@@ -0,0 +1,74 @@
+        sharedPrefs.unregisterOnSharedPreferenceChangeListener(preferenceChangeListener);
+        super.onInactive();
+    }
+}
```
missing new line at the end of the file

stefan-niedermann commented:
More changes since the last review:

- In case of an error, the `LiveData<SingleSignOnAccount>` no emits <code>null</code> instead of nothing
- Added `@WorkerThread` annotations to synchronous methods
- Aligned method names (`setCurrentSingleSignOnAccount` → `applyCurrentSingleSignOnAccount` / `commitCurrentSingleSignOnAccount`), backwards compatible
- Added some more documentation

stefan-niedermann commented:
Actually, reading from `SharedPreferences` *is* synchronously. It ain't "heavy", but it is. This is the reason why the [`DataStore`](https://developer.android.com/topic/libraries/architecture/datastore) framework (on top of `SharedPreferences`) promotes the async operations and explicitly warns about using the synchronous APIs.
